src/main.rs

```diff
@@ -402,5 +310,118 @@ fn create_manifest_from_config(
+        .into_report()
+        .attach_printable_lazy(|| {
+            format!(
+                "Failed to create generations file. Path: {}",
+                generations_path.display()
+            )
+        })
+        .change_context(MainError)?;
+
+    let bytes = to_bytes::<_, 1024>(&new_generations_file)
+        .into_report()
+        .attach_printable_lazy(|| "Failed to serialize generations file")
+        .change_context(MainError)?;
+
+    file.write_all(&bytes)
+        .into_report()
+        .attach_printable_lazy(|| {
+            format!(
+                "Failed to write to generations file. Path: {}",
+                generations_path.display()
+            )
+        })
+        .change_context(MainError)?;
+
+    let manifest = new_generations_file
+        .0
+        .into_iter()
+        .last()
+        .ok_or(MainError)
+        .into_report()
+        .attach_printable_lazy(|| "Failed to get latest manifest")
+        .change_context(MainError)?;
+
+    Ok(manifest.1)
+}
+
+type PackageWithSerializer = (Box<dyn SerializeLoaderInput>, WithSmith);
+
+fn resolve_package(
+    smiths: &[Box<dyn DynSmith>],
+    package: WithSmith,
+) -> Result<Vec<PackageWithSerializer>, MainError> {
+    let mut deps = package.package.package.dependencies.as_ref();
+    let none_map = BTreeMap::new();
+    if deps.is_none() {
+        deps = Some(&none_map);
+    }
+
+    let mut deps = deps
+        .unwrap()
+        .par_iter()
+        .map(|dep| {
+            let pkg = Package {
+                name: dep.0.clone(),
+                package: dep.1.clone(),
+            };
+
+            let smith_to_use = smiths
+                .iter()
+                .find(|s| s.get_package_name(&pkg.name).is_some())
+                .ok_or(MainError)
+                .into_report()
+                .attach_printable_lazy(|| {
+                    format!("Failed to find smith. Package name: {}", pkg.name)
+                })?;
+
+            let package = WithSmith {
+                smith: smith_to_use.name(),
+                package: pkg,
+            };
+
+            resolve_package(smiths, package)
+        })
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .flatten()
+        .collect::<Vec<_>>();
+
+    let smith_to_use = smiths
+        .iter()
+        .find(|s| s.name() == package.smith)
+        .ok_or(MainError)
+        .into_report()
+        .attach_printable_lazy(|| format!("Failed to find smith. Smith name: {}", package.smith))?;
+
+    let loader_data = smith_to_use
+        .resolve_dyn(&package.package)
+        .attach_printable_lazy(|| {
+            format!(
+                "Failed to resolve package. Package name: {}",
+                package.package.name
+            )
+        })
+        .change_context(MainError)?;
+
+    let final_package = (loader_data, package);
+    let mut final_vec = vec![final_package];
+    final_vec.append(&mut deps);
+
+    Ok(final_vec)
```

Comment:
Not stopping the review, but I'd prefer move this logic into its own file and keep the main file pretty small

Reply:
Sure I'll do that